parser-core/src/main/core/WidgetXMLLoader.scala

```diff
@@ -0,0 +1,379 @@
+                   )
+
+              val kids =
+                Seq( XMLElement( "setup", Map(), pen. setupCode, Seq())
+                   , XMLElement("update", Map(), pen.updateCode, Seq())
+                   )
+
+              XMLElement("pen", attrs, "", kids)
+
+          }
+
+        XMLElement("plot", attributes, "", children ++ pens)
+
+      case chooser: Chooser =>
+
+        val attributes =
+          Map( "left"    -> chooser.left.toString
+             , "top"     -> chooser.top.toString
+             , "right"   -> chooser.right.toString
+             , "bottom"  -> chooser.bottom.toString
+             , "current" -> chooser.currentChoice.toString
+             ) ++
+          ifDefined(chooser)( "display",  _.display) ++
+          ifDefined(chooser)("variable", _.variable)
+
+        val children =
+          chooser.choices.map {
+            case ChooseableString(string) =>
+              XMLElement("choice", Map("type" ->  "string", "value" ->           string), "", Seq())
+            case ChooseableDouble(double) =>
+              XMLElement("choice", Map("type" ->  "double", "value" ->  double.toString), "", Seq())
+            case ChooseableBoolean(boolean) =>
+              XMLElement("choice", Map("type" -> "boolean", "value" -> boolean.toString), "", Seq())
+            case ChooseableList(list) =>
+              val children = list.map((x) => XMLElement("value", Map("value" -> x.toString), "", Seq()))
+              XMLElement("choice", Map("type" -> "list"), "", children)
+          }
+
+        XMLElement("chooser", attributes, "", children)
+
+      case output: Output =>
+        val attributes =
+          Map( "left"     -> output.left.toString
+             , "top"      -> output.top.toString
+             , "right"    -> output.right.toString
+             , "bottom"   -> output.bottom.toString
+             , "fontSize" -> output.fontSize.toString
+             )
+        XMLElement("output", attributes, "", Seq())
+
+      case input: InputBox =>
+
+        val typeName =
+          input.boxedValue match {
+            case NumericInput(_, label) =>
+              label match {
+                case NumericInput.NumberLabel => "number"
+                case NumericInput.ColorLabel  => "color"
+              }
+            case input: StringInput =>
+              input.label match {
+                case StringInput.StringLabel   => "string"
+                case StringInput.ReporterLabel => "reporter"
+                case StringInput.CommandLabel  => "command"
+              }
+          }
+
+        val attributes =
+          Map( "left"      -> input.left.toString
+             , "top"       -> input.top.toString
+             , "right"     -> input.right.toString
+             , "bottom"    -> input.bottom.toString
+             , "multiline" -> input.multiline.toString
+             , "type"      -> typeName
+             ) ++
+          ifDefined(input)("variable", _.variable)
```

Comment:
How can this field (`variable`) ever be undefined?  I get that they might leave the textbox blank, but you can't have a valid (i.e. functioning) Input widget without it being tied to a global variable.  The same goes for the Switch, Slider, and Chooser widget-writers in this file.

Reply:
I saw that it was an Option so I assumed that it would be safer to check if it's None so the parser doesn't crash randomly. If it's definitely the case that it would never be None, this can certainly be removed, I just wasn't aware that was an invariant.

Reply:
Actually, upon further investigation, I discovered that the variable could actually be None, in the case of a HubNet widget. It seems like you can place a widget in the HubNet client editor without assigning a variable to it, but correct me if I'm wrong about that.

Reply:
RIP :skull_and_crossbones: 

Good catch.